vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java

```diff
@@ -0,0 +1,322 @@
+import io.vertx.core.http.ServerWebSocket;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.graphql.ApolloWSContext;
+import io.vertx.ext.web.handler.graphql.ApolloWSHandler;
+import io.vertx.ext.web.handler.graphql.ApolloWSOptions;
+import org.dataloader.DataLoaderRegistry;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+/**
+ * @author Rogelio Orts
+ */
+public class ApolloWSHandlerImpl implements ApolloWSHandler {
+
+  private static final Function<ApolloWSContext, Object> DEFAULT_QUERY_CONTEXT_FACTORY = rc -> rc;
+  private static final Function<ApolloWSContext, DataLoaderRegistry> DEFAULT_DATA_LOADER_REGISTRY_FACTORY = rc -> null;
+
+  private final static String HEADER_CONNECTION_UPGRADE_VALUE = "upgrade";
+
+  private final GraphQL graphQL;
+
+  private Function<ApolloWSContext, Object> queryContextFactory = DEFAULT_QUERY_CONTEXT_FACTORY;
+
+  private Function<ApolloWSContext, DataLoaderRegistry> dataLoaderRegistryFactory = DEFAULT_DATA_LOADER_REGISTRY_FACTORY;
+
+  private Handler<ServerWebSocket> connectionHandler;
+
+  private Handler<ServerWebSocket> endHandler;
+
+  private Handler<ApolloWSContext> messageHandler;
+
+  private ApolloWSOptions options;
+
+  public ApolloWSHandlerImpl(GraphQL graphQL, ApolloWSOptions options) {
+    Objects.requireNonNull(graphQL, "graphQL");
+    Objects.requireNonNull(options, "options");
+    this.graphQL = graphQL;
+    this.options = options;
+  }
+
+  @Override
+  public synchronized ApolloWSHandler queryContext(Function<ApolloWSContext, Object> factory) {
+    queryContextFactory = factory != null ? factory : DEFAULT_QUERY_CONTEXT_FACTORY;
+    return this;
+  }
+
+  @Override
+  public synchronized ApolloWSHandler dataLoaderRegistry(Function<ApolloWSContext, DataLoaderRegistry> factory) {
+    dataLoaderRegistryFactory = factory != null ? factory : DEFAULT_DATA_LOADER_REGISTRY_FACTORY;
+    return this;
+  }
+
+  @Override
+  public ApolloWSHandler connectionHandler(Handler<ServerWebSocket> connectionHandler) {
+    this.connectionHandler = connectionHandler;
+
+    return this;
+  }
+
+  @Override
+  public ApolloWSHandler endHandler(Handler<ServerWebSocket> endHandler) {
+    this.endHandler = endHandler;
+
+    return this;
+  }
+
+  @Override
+  public ApolloWSHandler messageHandler(Handler<ApolloWSContext> messageHandler) {
+    this.messageHandler = messageHandler;
+
+    return this;
+  }
+
+  @Override
+  public void handle(RoutingContext routingContext) {
+    MultiMap headers = routingContext.request().headers();
+    if(
+      headers.contains(HttpHeaders.CONNECTION)
+        &&
+      HEADER_CONNECTION_UPGRADE_VALUE.equals(headers.get(HttpHeaders.CONNECTION).toLowerCase())
```

Comment:
Protocol specified that on Connection header multiple values might exists along with upgrade. This should change to includes instead of equals

Reply:
@zenios good catch

@rogelio-o please do not update the PR, I'm merging it. I will update this part of the code.